ostd/src/logger.rs

```diff
@@ -2,36 +2,89 @@
 
 //! Logging support.
 
-use log::{Level, Metadata, Record};
+use alloc::format;
 
-use crate::early_println;
+use log::{LevelFilter, Metadata, Record};
 
-const LOGGER: Logger = Logger {};
+use crate::{
+    arch::timer::Jiffies,
+    boot::{kcmdline::ModuleArg, kernel_cmdline},
+    early_println,
+};
 
-/// The log level.
-///
-/// FIXME: The logs should be able to be read from files in the userspace,
-/// and the log level should be configurable.
-pub const INIT_LOG_LEVEL: Level = Level::Error;
+const LOGGER: Logger = Logger {};
 
 struct Logger {}
 
 impl log::Log for Logger {
     fn enabled(&self, metadata: &Metadata) -> bool {
-        metadata.level() <= INIT_LOG_LEVEL
+        metadata.level() <= log::max_level()
     }
 
     fn log(&self, record: &Record) {
-        if self.enabled(record.metadata()) {
-            early_println!("[{}]: {}", record.level(), record.args());
+        if !self.enabled(record.metadata()) {
+            return;
         }
+
+        let timestamp = format!("[{:>10?}]", Jiffies::elapsed().as_duration().as_secs_f64());
+        let level = format!("{:<5}", record.level());
+        let record_str = format!("{}", record.args());
+
+        #[cfg(feature = "log_color")]
+        let (timestamp, level, record_str) = {
+            use alloc::string::ToString;
+
+            use owo_colors::OwoColorize;
+
+            let timestamp = timestamp.green();
+            let level = match record.level() {
+                log::Level::Error => level.red().to_string(),
+                log::Level::Warn => level.bright_yellow().to_string(),
+                log::Level::Info => level.blue().to_string(),
+                log::Level::Debug => level.bright_green().to_string(),
+                log::Level::Trace => level.bright_black().to_string(),
+            };
+            let record_str = record_str.default_color();
+            (timestamp, level, record_str)
+        };
+
+        early_println!("{} {}: {}", timestamp, level, record_str);
     }
 
     fn flush(&self) {}
 }
 
+/// Initialize the logger. Users should avoid using the log macros before this function is called.
 pub(crate) fn init() {
-    log::set_logger(&LOGGER)
-        .map(|()| log::set_max_level(INIT_LOG_LEVEL.to_level_filter()))
-        .unwrap();
+    let level = match log_level_from_kcmd() {
+        Some(level) => level,
+        None => LevelFilter::Off,
+    };
+
+    log::set_max_level(level);
+    log::set_logger(&LOGGER).unwrap();
+}
+
+fn log_level_from_kcmd() -> Option<LevelFilter> {
```

Comment:
I have repeated myself many times and I am going to do it again: **The names of Rust functions (except getters and setters) should begin with _verbs_**. This function is not a simple getter; it involves a for loop to find the right configuration.